multipledispatch/utils.py

```diff
@@ -14,15 +19,22 @@ def expand_tuples(L):
 
     >>> expand_tuples([1, 2])
     [(1, 2)]
+
+    >>> expand_tuples([1, typing.Optional[str]])  #doctest: +ELLIPSIS
+    [(1, <... 'str'>), (1, <... 'NoneType'>)]
     """
     if not L:
         return [()]
-    elif not isinstance(L[0], tuple):
-        rest = expand_tuples(L[1:])
-        return [(L[0],) + t for t in rest]
     else:
-        rest = expand_tuples(L[1:])
-        return [(item,) + t for t in rest for item in L[0]]
+        if pytypes.is_Union(L[0]):
+           rest = expand_tuples(L[1:])
+           return [(item,) + t for t in rest for item in pytypes.get_Union_params(L[0])]
+        elif not pytypes.is_of_type(L[0], tuple):
```

Comment:
somehow this doesn't actually get hit when `L[0]` is not a tuple:

```
(Pdb) p L[0]
<type 'numpy.dtype'>
(Pdb) p pytypes.is_of_type(L[0], tuple)
True
```

This breaks importing `datashape`.

Reply:
The root cause of this is actually more concerning than this one bug. Why doesn't `pytypes.is_of_type` work correctly for type objects?